src/Features/Core/Portable/MakeFieldReadonly/MakeFieldReadonlyDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,190 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CodeStyle;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Operations;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.MakeFieldReadonly
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    internal sealed class MakeFieldReadonlyDiagnosticAnalyzer
+        : AbstractCodeStyleDiagnosticAnalyzer
+    {
+        public MakeFieldReadonlyDiagnosticAnalyzer()
+            : base(
+                IDEDiagnosticIds.MakeFieldReadonlyDiagnosticId,
+                new LocalizableResourceString(nameof(FeaturesResources.Add_readonly_modifier), FeaturesResources.ResourceManager, typeof(FeaturesResources)),
+                new LocalizableResourceString(nameof(FeaturesResources.Make_field_readonly), WorkspacesResources.ResourceManager, typeof(WorkspacesResources)))
+        {
+        }
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.ProjectAnalysis;
+
+        protected override void InitializeWorker(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(compilationStartContext =>
+            {
+                bool IsCandidateField(IFieldSymbol symbol) =>
+                        symbol.DeclaredAccessibility == Accessibility.Private &&
+                        !symbol.IsReadOnly &&
+                        !symbol.IsConst &&
+                        !symbol.IsImplicitlyDeclared &&
+                        symbol.Locations.Length == 1 &&
+                        !IsMutableValueType(symbol.Type);
+
+                // Stores the state for fields:
+                //  'isCandidate' : Indicates whether the field is a candidate to be made readonly based on it's declaration and options.
+                //  'written'     : Indicates if there are any writes to the field outside the constructor and field initializer.
+                var fieldStateMap = new ConcurrentDictionary<IFieldSymbol, (bool isCandidate, bool written)>();
+
+                void OnFieldWrite(IFieldSymbol field)
```

Comment:
can you name local functions "camelCase"?  Also, in the IDE we generally like to put hte local functions *after* the last return statement, with a "// local functions" comment on them.  It makes it really easy to read the normal function logic without having to jump over large nested functions in between.

Reply:
PascalCase is expected for local functions, to distinguish them from instances of delegates.

Reply:
IDE goes the opposite route in many cases.  It's camelCased (like all locals) since it is a local.  It also helps distinguish from actual methods.

Reply:
➡️ All instances of local functions in the code base current marked with a `// local functions` comment (case insensitive) use PascalCase. In the event of non-uniformity, PascalCase is preferred.

Reply:
Alright.  The naming is less of a concern for me and i'm fine with pascal being used.  The more important part is placing the code not inline with the main sequential logic for the function.  If we can move to after teh final return, that would be nice.  Thanks!

Reply:
I have switched back to using PascalCase naming for local functions. However, it is extremely frustrating that differing personal style preferences for different reviewers and different parts of Roslyn code base (IDE, compiler), are attempted to be enforced via code reviews:

1. `var` or explicit type: Reviewers on IDE team prefer var and those on compiler team prefer explicit type.
2. Local function names: Some IDE team reviewers prefer pascal casing, while some prefer camel casing. Reviewers on compiler team prefer camel casing.

Neither compiler nor IDE code bases have consistent usage in these style preferences to enforce this strongly via analyzers/editor config. This leads to the PR author to choose their own preferred style and waste time in iterations to jump between preferences as it gets reviewed.

I think the rule should be simple for all teams (compiler/IDE) wrt corresponding code bases they own: either make your code base consistent enough so you can enforce preferred style on your code base via editor config OR if this is too much work, then loosen the style requirements for PRs from folks that don't normally work on your code base and ensure that your team members follow these style preferences and cleanup code to abide by them whenever they change it.

Reply:
> Neither compiler nor IDE code bases have consistent usage in these style preferences to enforce this strongly via analyzers/editor config.

I disagree.  IDE is highly consistent with usage of "var".  I'm going to guess that it's 95% plus (and could be made more consistent pretty easily).

It's true that we're inconsistent with local function names.  I was hesitant originally around allowing local functions, but we let them creep in slowly without doing a team assessment about what we'd like the pattern there to be.  Team should decide acceptable style (probably whatever the majority of the IDE code uses), and then apply that consistently.

Reply:
> This leads to the PR author to choose their own preferred style and waste time in iterations to jump between preferences as it gets reviewed.

The IDE has strongly been opinionated on 'var' for over a decade (and many other non-checked systems).  PR authors should respect that.  IDE shoudl also attempt to see what can be enforced stylistically, potentially through tools like StyleCopAnalyzers.  I know @sharwell seems particularly interested in this, and i would be totally supportive of adopting such a system on the *IDE* side.

Note that the compiler is its own project with its own rules.  Devs are expected to know and follow the conventions each team prefers for their respective codebases.

Reply:
> I think the rule should be simple for all teams (compiler/IDE) wrt corresponding code bases they own: either make your code base consistent enough so you can enforce preferred style on your code base via editor config

Yes.  I would be fine with this.  @sharwell For the IDE team, do you know why we haven't flipped on certain options?

> then loosen the style requirements for PRs from folks that don't normally work on your code base 

Sorry, i don't think that's ok.  There are thousands of projects out there that accept contributions from external people, but expect those people to follow the conventions of that project.  That ranges from coding style, to tests, to documentation, to whatever.  The onus is on the contributer to follow the wishes of the project they're contributing to.  Though i certainly agree anything the project can do to make that easier is desirable.  However, the lack of nice-ness here is not an pass to just say that contributers don't have to care and the team is responsible for cleaning up after the fact.